src/pip/_internal/download.py

```diff
@@ -686,7 +686,8 @@ def iter_secure_origins(self):
         for secure_origin in SECURE_ORIGINS:
             yield secure_origin
         for host in self.pip_trusted_hosts:
-            yield ('*', host, '*')
+            parsed = urllib_parse.urlparse(build_url_from_netloc(host))
```

Comment:
This will introduce extra complexity to convert parsed (host, port) pairs to string type hosts for the finder to use.

Reply:
It just reflects the values passed from `--trusted-host` option, I am OK with the current name.

Reply:
It's better to parse the value once at the beginning instead of every time it's used. For the name, the reason to use a different name is to reflect that this value will be pairs, while `trusted_host` is strings.

Reply:
Sounds reasonable, so can I keep both attributes to save the efforts of conversion?

Reply:
Okay, thanks.

> so can I keep both attributes to save the efforts of conversion?

Can you clarify this question?

(By the way, one more point is that we're already parsing the `trusted_host` value anyways inside `add_trusted_host()`. That takes place in `netloc_has_port()`, so you might want to replace `netloc_has_port()` with a `parse_netloc()` function that returns the host-port pair.)

Reply:
That means we should keep the original strings(`pip_trusted_hosts`) instead of building them back from the parsed (host, port) pairs(`pip_trusted_origins`). With that said, we will have two attributes.

Refactoring `netloc_has_port()` is a good idea, I will take it.

Reply:
I was thinking not to keep the original strings so there would be one source of truth. It should be a trivial thing to concatenate the host and port if necessary (and in particular easier than parsing). Converting from host-port strings to structured tuple is analogous to converting between bytes and unicode at process boundaries. In this case, the step of going back to strings is needed for doing a subprocess call with a `--trusted-host` argument.

Reply:
The updated `PackageFinder` property can look something like--

```python
@property
def trusted_hosts(self):
    # type: () -> Iterable[str]
    host_ports = self._link_collector.session.pip_trusted_origins
    yield from (build_netloc(*host_port) for host_port in host_ports)
```

where `build_netloc()` is the function that does the concatenation and has signature `build_netloc(host, port)`.

Reply:
I came across another thought: what if the index URL is HTTPS and contains auth part?

`pip install -i https://USER:PASSWD@example.com --trusted-host example.com`.

with the current implementation, this will fail because the index URL is not caught by the insecure session adapter.

Reply:
Is this point about this PR or separate from the PR?

Reply:
NVM, it should be another improvement.

Reply:
You can file it as a separate issue if warranted..